crypto/x509/x509_vfy.c

```diff
@@ -1734,22 +1735,27 @@ static int internal_verify(X509_STORE_CTX *ctx)
      * is allowed to reset errors (at its own peril).
      */
     while (n >= 0) {
-        EVP_PKEY *pkey;
-
         /*
-         * Skip signature check for self signed certificates unless explicitly
-         * asked for.  It doesn't add any security and just wastes time.  If
-         * the issuer's public key is unusable, report the issuer certificate
+         * Skip signature check for self-signed certificates unless explicitly
+         * asked for because it does not add any security and just wastes time.
+         * If the issuer's public key is not available or its key usage does
+         * not support issuing the subject cert, report the issuer certificate
          * and its depth (rather than the depth of the subject).
          */
         if (xs != xi || (ctx->param->flags & X509_V_FLAG_CHECK_SS_SIGNATURE)) {
+            EVP_PKEY *pkey;
+            int issuer_depth = n + (xi == xs ? 0 : 1);
+            int ret = X509_signing_allowed(xi, xs);
+
+            if (ret != X509_V_OK && !verify_cb_cert(ctx, xi, issuer_depth, ret))
+                return 0;
             if ((pkey = X509_get0_pubkey(xi)) == NULL) {
-                if (!verify_cb_cert(ctx, xi, xi != xs ? n+1 : n,
-                        X509_V_ERR_UNABLE_TO_DECODE_ISSUER_PUBLIC_KEY))
+                ret = X509_V_ERR_UNABLE_TO_DECODE_ISSUER_PUBLIC_KEY;
+                if (!verify_cb_cert(ctx, xi, issuer_depth, ret))
```

Comment:
I'm not sure what the point of this change is, assigning that value to ret does not seem to have any effect.

Reply:
The value is used (as before) as argument of `verify_cb_cert`.
I assigned it to `ret` just to remove the line break in the call of `verify_cb_cert`, making use of the variable that was there already anyway.

Reply:
I could of course change this back to
```
                if (!verify_cb_cert(ctx, xi, issuer_depth,
                                    X509_V_ERR_UNABLE_TO_DECODE_ISSUER_PUBLIC_KEY))
```
but then the line would be more than 80 characters wide.

Reply:
The primary style rule is "be readable"
```
; g ls-files | grep '.[ch]$' | xargs awk '{if (length($0)>80) print;}' | wc -l
3845
```

Reply:
Fine, so I reverted this change (line is now 83 chars long)
and did the analogous change also a few lines later.